main/src/library/Vector.flix

```diff
@@ -0,0 +1,1389 @@
+            } else {
+                let x = get(i - 1, v);
+                loop(i - 1, x :: acc)
+            }
+        };
+        loop(length(v), Nil)
+
+    ///
+    /// Optionally returns the vector `v` as a non-empty list.
+    ///
+    /// If `v` is empty return `None`, otherwise return the Nel wrapped in `Some`.
+    ///
+    pub def toNel(v: Vector[a]): Option[Nel[a]] =
+        def loop(i, acc) = {
+            if (i == 0) {
+                acc
+            } else {
+                let x = get(i - 1, v);
+                loop(i - 1, Nel.cons(x, acc))
+            }
+        };
+        if (Array.isEmpty(v)) {
+            None
+        } else {
+            let i = length(v) - 1;
+            Some(loop(i, Nel.singleton(get(i, v))))
+        }
+
+    ///
+    /// Returns the vector `v` as a `DelayList`.
+    ///
+    pub def toDelayList(v: Vector[a]): DelayList[a] =
+        def loop(i, acc) = {
+            if (i < 0)
+                acc
+            else
+                loop(i - 1, DelayList.ECons(get(i, v), acc))
+        };
+        loop(length(v) - 1, DelayList.ENil)
+
+    ///
+    /// Returns the vector `v` as a chain.
+    ///
+    pub def toChain(v: Vector[a]): Chain[a] =
+        def loop(i, acc) = {
+            if (i == 0) {
+                acc
+            } else {
+                let x = get(i - 1, v);
+                loop(i - 1, Chain.cons(x, acc))
+            }
+        };
+        loop(length(v), Chain.empty())
+
+    ///
+    /// Optionally returns the vector `v` as a non-empty chain.
+    ///
+    /// If `v` is empty return `None`, otherwise return the Nec wrapped in `Some`.
+    ///
+    pub def toNec(v: Vector[a]): Option[Nec[a]] =
+        def loop(i, acc) = {
+            if (i == 0) {
+                acc
+            } else {
+                let x = get(i - 1, v);
+                loop(i - 1, Nec.cons(x, acc))
+            }
+        };
+        if (isEmpty(v)) {
+            None
+        } else {
+            let i = length(v) - 1;
+            Some(loop(i, Nec.singleton(get(i, v))))
+        }
+
+    ///
+    /// Returns `v` as a MutDeque.
+    ///
+    pub def toMutDeque(rc: Region[r], v: Vector[a]): MutDeque[a, r] \ Write(r)   =
+        let d = new MutDeque(rc);
+        forEach(x -> MutDeque.pushBack(x, d), v);
+        d
+
+    ///
+    /// Returns `Some(x)` if `x` is the first element of `v`.
+    ///
+    /// Returns `None` if `v` is empty.
+    ///
+    pub def head(v: Vector[a]): Option[a] =
+        nth(0, v)
+
+    ///
+    /// Returns `Some(x)` if `x` is the last element of `v`.
+    ///
+    /// Returns `None` if `v` is empty.
+    ///
+    pub def last(v: Vector[a]): Option[a] =
+        nth(length(v) - 1, v)
+
+    ///
+    /// Return a new vector, appending the elements `b` to elements of `a`.
+    ///
+    pub def append(a: Vector[a], b: Vector[a]): Vector[a] = region rc {
+        buildVector(rc, length(a) + length(b), arr -> {
+            Array.updateSequence!(0, a, arr);
```

Comment:
This would need to become some more clever use of buildVector reading from a and then from b.

Reply:
`buildVector` is now eliminated. `arrayUpdateSeqV!` is used to update the interim array without needing to change `a` and `b` into arrays or cheating (i.e. relying on them being a type alias to array).

[`a` and `b` are now called `v1` and v2` to emphasize they are vectors]